plutus-metatheory/src/Raw.hs

```diff
@@ -63,34 +66,41 @@ unIndex (Index n) = toInteger n
 convP :: Program NamedTyDeBruijn NamedDeBruijn DefaultUni DefaultFun a -> RTerm
 convP (Program _ _ t) = conv t
 
+convK :: Kind a -> KIND
+convK (Type _)           = Star
+convK (KindArrow _ k k') = Arrow (convK k) (convK k')
+
 convT :: Type NamedTyDeBruijn DefaultUni a -> RType
 convT (TyVar _ (NamedTyDeBruijn x)) = RTyVar (unIndex (ndbnIndex x))
 convT (TyFun _ _A _B)               = RTyFun (convT _A) (convT _B)
-convT (TyForall _ _ _K _A)          = RTyPi (() <$ _K) (convT _A)
-convT (TyLam _ _ _K _A)             = RTyLambda (() <$ _K) (convT _A)
+convT (TyForall _ _ _K _A)          = RTyPi (convK _K) (convT _A)
+convT (TyLam _ _ _K _A)             = RTyLambda (convK _K) (convT _A)
 convT (TyApp _ _A _B)               = RTyApp (convT _A) (convT _B)
-convT (TyBuiltin _ b)               = RTyCon (convTyCon b)
+convT (TyBuiltin _ b)               = convTyCon b
```

Comment:
This may fail, given `error "unsupported builtin type application"` below. You should handle the `DefaultUniApply` case (by elaborating it to `RTyApp`) and only call `convTyCon` if `b` is not a type application.

Reply:
Aren't the good cases covered by lines 95-98?
```
convTyCon (SomeTypeIn (DefaultUniList a))             =
    RTyApp (RTyCon RTyConList) (convTyCon (SomeTypeIn a))
convTyCon (SomeTypeIn (DefaultUniPair a b))           =
    RTyApp (RTyApp (RTyCon RTyConPair) (convTyCon (SomeTypeIn a))) (convTyCon (SomeTypeIn b))
```

Reply:
:woman_facepalming: I have no idea how I missed that, sorry!

Anyways, those lines aren't quite correct. One can have a partially applied `DefaultUniProtoPair` for example and then neither of these cases would catch it. Just elaborate any `DefaultUniApply` to `RTyApp` and handle `DefaultUniProtoList` and `DefaultUniProtoPair` to handle all possible cases and you'll be able to avoid throwing that `error` there too.

Also I'd define `convTyCon` over `DefaultUni (Esc a)` rather than `SomeTypeIn DefaultUni`, so that you don't need to match on `SomeTypeIn` repeatedly.

Reply:
Thank you Roman.

> Also I'd define convTyCon over DefaultUni (Esc a) rather than SomeTypeIn DefaultUni, so that you don't need to match on SomeTypeIn repeatedly.

You mean define `convTyCon :: DefaultUni (Esc a) -> RType`?
Wouldn't this open the existential prematurely andfail to typecheck?

Reply:
I mean fail to typecheck when calling it
```
convT (TyBuiltin _ (SomeTypeIn b))  = convTyCon b
```